### ISSUE TITLE
external sendmail command for mail notifier

### ISSUE THREAD

**Issue by tardyp** [buildbot/buildbot#2922]
This ticket is a migrated Trac ticket [3196](http://trac.buildbot.net/ticket/3196)

People contributed to the original ticket: @unknown_contributor, @djmitche
Ticket created on: `Feb 10 2015`
Ticket last modified on: `Feb 11 2015`

---

MailNotifier works via twisted-mail for now, which sometimes leads to trouble.
One problem is absence of mail queue in twisted. Restarting buildbot process leads to all queued mail lost.
Another problem is TLS/SSL support, which may be very different for all SMTP servers out there. For example, STARTTLS is required by twisted to work over SSL, but many servers offer separate port, awaiting for pure SSL connection to it. Some servers may block authenticated connections over 25 port, so twisted will fail sending email, while reqular mail user agent, such as Outlook will work.

This problems can be addressed by supporting mail delivery via system '/usr/bin/sendmail' command which is a well-known way to send email in unix world.
Mail transfer agents like postfix and exim provide compatible /usr/bin/sendmail replacement, allowing local mail be sent with no authorization and delivered based on configuration, maintaining message queues, mail routing and so on.
There are some programs like msmtp or ssmtp, able to talk to any smtp server, supporting TLS and authorization of all kinds.

It would be nice to have ability for buildbot's [[MailNotifier]] work with local sendmail command, as it's simple to implement and leads to great flexibility. This also let's to avoid installing additional python ssl modules (which can be triky on older enterprise systems).
This also makes error reporting better, as twisted-mail with no ssl claims that server not supporting TLS, which is very confusing for peaple not familiar with pyhon and twisted.

---

**Comment by seankelly** [comment]
> Another problem is TLS/SSL support, which may be very different for all SMTP servers out there. For example, STARTTLS is required by twisted to work over SSL, but many servers offer separate port, awaiting for pure SSL connection to it.

There is now an `useSmtps` argument for this part, added in #2495.

**Comment by kfriedberger** [comment]
What is the status of this ticket?
Can I (or "How can I") configure the buildbot-master's mail-notifier to send mails via local "sendmail"?

**Comment by pbrkr** [comment]
We're using sSMTP as a lightweight way of sending emails from our servers so we need to be able to send emails via /usr/bin/sendmail. I think this issue needs to be revived as a current feature request.

What would need to change to support this? I may be able to implement it with some guidance.

**Comment by p12tic** [comment]
@pbrkr: I agree that the feature makes sense. I could guide you through implementing this.

@tardyp: Do we want to have a separate class for "external" mail notifier or to put everyting to MailNotifier and then select the backend via a string parameter? I don't like both of the options because they'll still result in a fair amount of duplication (even if it's just transferring 20 parameters to a common mixin).
